### PR TITLE
[CoAP Server] Message deduplication

### PR DESCRIPTION
**Pull request by hubertmis**
Implementation of RFC 7252 4.5.

### CODE REVIEW

File: src/core/coap/coap_server.cpp [hunk 1]
```diff
@@ -204,5 +228,174 @@ ThreadError Server::SetPort(uint16_t aPort)
...
+    ThreadError            error = kThreadError_NotFound;
+    Message               *message;
+    EnqueuedResponseHeader enqueuedResponseHeader;
+    Ip6::MessageInfo       messageInfo;
+    Header                 header;
+
+    for (message = mQueue.GetHead(); message != NULL; message = message->GetNext())
+    {
+        enqueuedResponseHeader.ReadFrom(*message);
+        memcpy(&messageInfo, &enqueuedResponseHeader.GetMessageInfo(), sizeof(messageInfo));
```
Comment (jwhui) [inline src/core/coap/coap_server.cpp]: nit: use assignment operator instead?

File: src/core/coap/coap_server.cpp [hunk 2]
```diff
@@ -204,5 +228,174 @@ ThreadError Server::SetPort(uint16_t aPort)
...
+
+        // Check source endpoint
+        if (messageInfo.GetPeerPort() != aMessageInfo.GetPeerPort())
+        {
+            continue;
+        }
+
+        if (memcmp(&messageInfo.GetPeerAddr(), &aMessageInfo.GetPeerAddr(), sizeof(Ip6::Address)))
```
Comment (jwhui) [inline src/core/coap/coap_server.cpp]: nit: use `==` operator instead?

File: src/core/coap/coap_server.hpp [hunk 3]
```diff
@@ -94,6 +96,193 @@ class Resource : public otCoapResource
...
+        return aMessage.Read(aMessage.GetLength() - sizeof(*this), sizeof(*this), this);
+    }
+
+    /**
+     * This method removes metadata from the message.
+     *
+     * @param[in]  aMessage  A reference to the message.
+     *
+     * @retval kThreadError_None    Successfully removed the bytes.
+     * @retval kThreadError_NoBufs  Insufficient available buffers to place the message.
```
Comment (jwhui) [inline src/core/coap/coap_server.hpp]: Is `kThreadError_NoBufs` possible?

File: src/core/coap/coap_server.cpp (hunk 1 above)
Comment (hubertmis) [inline src/core/coap/coap_server.cpp]: :+1:

File: src/core/coap/coap_server.cpp (hunk 2 above)
Comment (hubertmis) [inline src/core/coap/coap_server.cpp]: :+1:

File: src/core/coap/coap_server.hpp (hunk 3 above)
Comment (hubertmis) [inline src/core/coap/coap_server.hpp]: Called function returns `_NoBufs`. I followed the code and it seems impossible in this case. I'll assert it and change function type to void.

**Pull request merged**